README.md

```diff
@@ -4,128 +4,361 @@ Daniel Ehrenberg, Igalia. Stage 3
+### Rounding
 
-Many (all?) other dynamically typed programming languages which have multiple numeric types implement a *numeric tower*. This forms an ordering between types--on the built-in numeric types, when an operator is used with operands from two types, the greater type is chosen as the domain, and the "less general" operand is cast to the "more general" type. Unfortunately, as the previous example shows, there is no "more general" type between arbitrary integers and double-precision floats. The typical resolution, then, is to take floats as the "more general" type.
+As noted above, the `BigInt` only represents whole numbers. `Number` only reliably represents integers up to 2<sup>53</sup>. That means both dividing and converting to a `Number` can lead to rounding.
 
-Silently losing precision sometimes may be a problem, but in most dynamically typed programming languages which provide integers and floats, integers are written like `1` and floats are written like `1.0`. It's possible to scan code for operations which may introduce floating point precision by looking for a decimal point. JavaScript exacerbates the scope of losing precision by making the unfortunate decision that a simple literal like `1` is a float. So, if mixed-precision were allowed, an innocent calculation such as `2n ** 53n + 1` would produce the float `2**53`--defeating the core functionality of this feature.
+```js
 
-To avoid this problem, this proposal bans implicit coercions between Numbers and BigInts, including operations which are mixed type. `1n + 1` throws a TypeError. So does passing `1n` as an argument into any JavaScript standard library function or Web API which expects a Number. Instead, to convert between types, an explicit call to `Number()` or `BigInt()` needs to be made to decide which domain to operate in. `0 === 0n` returns `false`.
+5n / 2n
+// ↪ 2n
 
-Comparisons form an exception to this rule: It's mathematically well-defined to allow comparison operators such as `<` and `==` compare between Numbers and BigInts. Unlike operators like `+`, there is no loss of precision, since the output is just a Boolean. Further, `==` is extended in comparisons with strings in analogous ways, such as `0n == ""` is `true`. Although we may not be able to extend this support to user-defined types, it seems sufficiently important for meeting user expectations that this proposal adds them as a one-off.
+Number(151851850485185185047n)
+// ↪ 151851850485185200000
+
+```
+
+### Other Exceptions
+
+Attempting to convert a fractional value to a `BigInt` throws an exception both when the value is represented as an `Number` and a `String`.
+
+```js
+BigInt(1.5)
+// ↪ RangeError: The number 1.5 is not a safe integer and thus cannot be converted to a BigInt
+
+BigInt('1.5')
+// ↪ SyntaxError: Cannot convert 1.5 to a BigInt
+
+```
+
+`Math.round`, `Math.fround` and `|` throw exceptions when called with a `BigInt` since these functions are defined as returning a `Number`.
```

Comment:
I updated this to remove the `asm.js` mention and add in `Math.round` as well, but when plen playing with the console, I ran into an interesting conundrum: `parseInt`and `parseFloat` both _do not_ throw an error, even though they come with the same possible loss of precision that motivates the TypeError discussed in #153 for exponentiation.

It is a bit difficult for me to pin down the model for intuition in this case. For `parseInt`, I understand sort of, in the sense that is an explicit coercion to a number, but so is, say, `Math.round`. I would love to be able to reflect a model that would help users know where to expect TypeErrors and where not to. Is there something I am missing?

If not, I wonder if this is worth opening an issue to chat about more widely, adding a comment to the exponent thread, or something else as well. What do you think is best?

Reply:
Note, parseInt and parseFloat ignore trailing garbage, so `parseInt("123n")` returns the Number 123, just like `parseInt("123khasdflkjahfda")` does. It would be difficult to change these functions due to web-compatibility concerns. I think the most relevant thing to say about them is, they are weird legacy functions that do a bunch of rounding/slicing/ignoring and output Numbers; they have no relation to BigInt. I don't think this has much bearing on the semantics of `Math.round`.

Happy to chat about this somehow, on whatever thread, but I don't anticipate that we'll make changes to the BigInt proposal at this point unless the issue is very serious, given how it's shipping in Chrome and Node.js.